### PR TITLE
[JENKINS-36532] General refactor

### PR DESCRIPTION
**Pull request by alexsomai**
Related to issue [[JENKINS-36532]](https://issues.jenkins-ci.org/browse/JENKINS-36532)

Summary of this pull request:
- [X] Include the changes suggested in #1 
- [X] Use _DataBoundSetter_ where makes sense, to minimize the output of _Snippet Generator_ code
- [X] Rename public exposed method API to `select(run, context)`
- [x] The parameters that are mandatory in the `RunSelectorContext` should be set in the constructor. Only optional parameters, like `verbose` should have setters 

CC @ikedam @oleg-nenashev @martinda

### CODE REVIEW

**Comment by ikedam** [comment]
> Remove ParameterizedRunSelector, since it supports only variable names, and not variable expressions

Sorry, I confused you.
`ParameterizedRunSelector` supports both variable names (FILTER) and variable expressions (${FILTER}).
You can deprecate and remove the variable name feature (FILTER) and have it support only variable expressions (${FILTER}).

**Comment by ikedam** [comment]
JFYI: `assertEqualDataBoundBeans` in jenkins-test-harness 2.12 also checks properties with `DataBoundSetter`.
You can switch to it by setting `jenkins-test-harness.version` in pom.xml (See https://github.com/jenkinsci/plugin-pom for details).

File: src/main/java/org/jenkinsci/plugins/runselector/RunSelector.java [hunk 1]
```diff
@@ -47,7 +47,7 @@
      * @throws IOException if an error occurs while performing the operation.
      * @throws InterruptedException if any thread interrupts the current thread.
      */
-    public Run<?, ?> pickBuildToCopyFrom(@Nonnull Job<?,?> job, @Nonnull final RunSelectorPickContext context)
+    public Run<?, ?> select(@Nonnull Job<?,?> job, @Nonnull final RunSelectorContext context)
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/runselector/RunSelector.java:52]: Good practice. Add annotations like CheckForNull in cases, where null results are the part of the common execution flow

File: src/main/java/org/jenkinsci/plugins/runselector/context/RunSelectorContext.java [hunk 2]
```diff
@@ -143,6 +148,58 @@ public boolean isVerbose() {
...
+    }
+
+    /**
+     * The project name to copy from.
+     * Be aware that this might be different from the full name of the project
+     * as it might be specified with relative expression.
+     * 
+     * @return the project name to copy from.
+     */
+    @Nonnull
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/runselector/context/RunSelectorContext.java]: I'm not convinced that it is never null. No guarantee setProjectName() gets invoked in API users, null can be also passed there

File: src/main/java/org/jenkinsci/plugins/runselector/context/RunSelectorPickContext.java [hunk 3]
```diff
@@ -1,109 +0,0 @@
...
-
-import javax.annotation.CheckForNull;
-import javax.annotation.Nonnull;
-
-/**
- * Context for the task of runselector picking the build to copy.
- * This allows us to adding new fields without affecting
- * existing plugins.
- */
-public class RunSelectorPickContext extends RunSelectorCommonContext {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/runselector/context/RunSelectorPickContext.java:39]: Do you use RunSelectorContext instead?

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/runselector/context/RunSelectorPickContext.java:39]: Yes

**Comment by oleg-nenashev** [comment]
I would vote for RunSelectorContext initialization refactoring. The current implementation seems to be a bit fragile

**Comment by alexsomai** [comment]
> You can deprecate and remove the variable name feature (FILTER) and have it support only variable expressions (${FILTER}).

I've dropped the commit in which I've completely removed the `ParameterizedRunSelector`.

But I'm not sure what changes I should make. In the `resolveParameter()` method, should I replace the following lines:

``` java
String xml = context.getEnvVars().get(getParameterName());
if (xml == null) {
    context.logInfo("{0} is not defined", getParameterName());
}
return xml;
```

with

``` java
return null;
```

?
I don't understand well why to remove this feature. People are not using it?

File: pom.xml [hunk 4]
```diff
@@ -1,168 +1,185 @@
...
+        <version>2.9</version>
     </parent>
 
-    <artifactId>run-selector</artifactId>
+    <artifactId>run-selector-api</artifactId>
     <packaging>hpi</packaging>
-    <name>Run Selector Plugin</name>
-    <version>1.0-SNAPSHOT</version>
-    <url>http://wiki.jenkins-ci.org/display/JENKINS/Run+Selector+Plugin</url>
+    <name>Run Selector API Plugin</name>
```
Comment (alexsomai) [inline pom.xml:13]: WDYT if we call it Run Selector **API** Plugin, because it's not a standalone plugin.

File: pom.xml [hunk 5]
```diff
@@ -1,168 +1,185 @@
...
+            <email>devld@ikedam.jp</email>
+        </developer>
+    </developers>
+
     <properties>
-      <jenkins.version>1.580.1</jenkins.version>
-      <java.level>6</java.level>
-      <workflow.version>1.4.2</workflow.version>
+        <jenkins.version>1.580.1</jenkins.version>
+        <java.level>7</java.level>
```
Comment (alexsomai) [inline pom.xml]: Should I revert back to java.level 6?

File: pom.xml [hunk 6]
```diff
@@ -1,168 +1,185 @@
...
     </licenses>
 
+    <developers>
+        <developer>
+            <id>alexsomai</id>
+            <name>Alexandru Somai</name>
+            <email>somai.alexandru@gmail.com</email>
+        </developer>
+        <developer>
+            <id>ikedam</id>
```
Comment (alexsomai) [inline pom.xml]: @ikedam do you mind if I add you as a maintainer too? Because you know better than me how the selectors work.

**Comment by oleg-nenashev** [comment]
Please use <code>@Restricted(NoExternalUse)</code> for DataBound setters, which are not supposed to be a part of public API.

File: pom.xml (hunk 5 above)
Comment (oleg-nenashev) [inline pom.xml]: If the target core is below 1.609.x, Java 6 should be supported

File: pom.xml (hunk 4 above)
Comment (oleg-nenashev) [inline pom.xml:13]: LGTM

Comment (ikedam) [inline pom.xml:13]: +1 for run-selector-api.

File: pom.xml (hunk 6 above)
Comment (ikedam) [inline pom.xml]: I don't mind adding me itself, but I believe it's better to have there only one maintainer to clarify who maintains this plugin.
Anyway, I don't think I can have much time to maintain this plugin.

File: pom.xml [hunk 7]
```diff
@@ -1,168 +1,185 @@
...
+    </developers>
+
     <properties>
-      <jenkins.version>1.580.1</jenkins.version>
-      <java.level>6</java.level>
-      <workflow.version>1.4.2</workflow.version>
+        <jenkins.version>1.580.1</jenkins.version>
+        <java.level>7</java.level>
+        <workflow.version>1.4.2</workflow.version>
+        <jenkins-test-harness.version>2.1</jenkins-test-harness.version>
```
Comment (ikedam) [inline pom.xml]: You don't need to specify jenkins-test-harness.version (at least for 2.1)

File: src/main/java/org/jenkinsci/plugins/runselector/context/RunSelectorContext.java [hunk 8]
```diff
@@ -314,19 +347,50 @@ private void copyExtensionListFrom(RunSelectorCommonContext src) {
...
+     * Constructs the environment variables for the current build.
+     *
+     * @return the current build environment variables
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private EnvVars constructEnvVars() throws IOException, InterruptedException {
+        EnvVars envVars = build.getEnvironment(listener);
+        if (build instanceof AbstractBuild) {
+            envVars.overrideAll(((AbstractBuild<?, ?>) build).getBuildVariables()); // Add in matrix axes..
```
Comment (ikedam) [inline src/main/java/org/jenkinsci/plugins/runselector/context/RunSelectorContext.java]: Please use `envVars.putAll` instead, as `overrideAll` removes empty variables.
Copyartifact has a bug for that: https://issues.jenkins-ci.org/browse/JENKINS-36554

**Comment by ikedam** [comment]
> I don't understand well why to remove this feature. People are not using it?
- The variable expression feature was introduced for pipilenes (aka. workflow): https://issues.jenkins-ci.org/browse/JENKINS-30357
  - The variable name feature remains for the backward compatibility.
- The varable expression feature can be alternate for the variable name feature.
- These two features can be conflicted.
  - VAR1=VAR2, VAR2=blahblahblah, and ${VAR1} will be expanded to blahblahblah.
- Confusing if there're two ways for same configurations.
- More tests are needed.
- `ParameterizedBuildFilter` supports only variable expressions.

And I think you can remove this feature.

On the other hand, you can continue to support this feature.
The variable expression feature is introduced recently, and many users of copyartifact may still use variable names.

> But I'm not sure what changes I should make. 

The way you wrote is right.

File: pom.xml (hunk 5 above)
Comment (ikedam) [inline pom.xml]: Why do you want to switch to java7?
Rather it would be better to switch to Jenkins >= 1.609 if you really need Java7.

**Comment by ikedam** [comment]
Looks good to me.

**Pull request merged**